src/main/kotlin/com/openlattice/codex/CodexService.kt

```diff
@@ -0,0 +1,187 @@
+package com.openlattice.codex
+
+import com.auth0.json.mgmt.users.User
+import com.google.common.collect.ArrayListMultimap
+import com.google.common.collect.ListMultimap
+import com.hazelcast.core.HazelcastInstance
+import com.hazelcast.core.IMap
+import com.openlattice.apps.AppConfigKey
+import com.openlattice.apps.AppTypeSetting
+import com.openlattice.codex.controllers.CodexConstants
+import com.openlattice.data.DataEdge
+import com.openlattice.data.DataGraphManager
+import com.openlattice.data.EntityDataKey
+import com.openlattice.data.EntityKeyIdService
+import com.openlattice.datastore.apps.services.AppService
+import com.openlattice.datastore.services.EdmManager
+import com.openlattice.edm.type.PropertyType
+import com.openlattice.hazelcast.HazelcastMap
+import com.openlattice.notifications.sms.PhoneNumberService
+import com.openlattice.organizations.roles.SecurePrincipalsManager
+import com.twilio.rest.api.v2010.account.Message
+import org.joda.time.DateTime
+import org.slf4j.LoggerFactory
+import java.time.Instant
+import java.time.OffsetDateTime
+import java.time.ZoneOffset.UTC
+import java.util.*
+import javax.servlet.http.HttpServletRequest
+
+class CodexService(
+        val hazelcast: HazelcastInstance,
+        val appService: AppService,
+        val edmManager: EdmManager,
+        val dataGraphManager: DataGraphManager,
+        val entityKeyIdService: EntityKeyIdService,
+        val principalsManager: SecurePrincipalsManager
+) {
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(CodexService::class.java)
+    }
+
+    val appId = appService.getApp(CodexConstants.APP_NAME).id
+    val typesByFqn = CodexConstants.AppType.values().associate { it to appService.getAppType(it.fqn) }
+    val appConfigs: IMap<AppConfigKey, AppTypeSetting> = HazelcastMap.APP_CONFIGS.getMap(hazelcast)
+    val propertyTypesByAppType = typesByFqn.values.associate { it.id to edmManager.getPropertyTypesOfEntityType(it.entityTypeId) }
+    val propertyTypesByFqn = propertyTypesByAppType.values.flatMap { it.values }.associate { it.type to it.id } // TODO check if collisions break this
+
+    fun processOutgoingMessage(message: Message, organizationId: UUID, senderId: String) {
+        val dateTime = formatDateTime(message.dateCreated)
+        val phoneNumber = message.to
+        val messageId = message.sid
+        val text = message.body
+
+        val sender = principalsManager.getUser(senderId)
+
+        /* create entities */
+
+        val contactEDK = getContactEntityDataKey(organizationId, phoneNumber)
+        val messageEDK = getMessageEntityDataKey(organizationId, dateTime, messageId, text)
+
+        /* create associations */
+
+        val associations: ListMultimap<UUID, DataEdge> = ArrayListMultimap.create()
```

Comment:
¯\_(ツ)_/¯ that's what dgm expects + what everything is currently passing it

We definitely needed to get rid of multimaps specifically for updateEntitiesInEntitySet since multimaps treat empty collection values as entry not present, and this presented problems for PartialReplace (because replacing a set of values with [] should effectively clear the existing values, but instead nothing was changed). I'm not sure what the reason for removing the other multimaps is.

Reply:
Looks like @you-leee updated her PRs to fix it in this instance already

Reply:
I also do not know the base reason but I prefer having as few types to think about as possible